# Issue 23785: Fix sagelib setup.py install issues on Python 3

archive/issues_023548.json:
```json
{
    "body": "This fixes a few issues, particularly with `clean_stale_files` on Python 3.  In particular, it prevents deletion of valid extension module and Python bytecode files, by better handling of `__pycache__` and the more complex filename extensions used in Python 3.\n\n**CC:**  @fchapoton\n\n**Branch/Commit:** [7cb6b3d19311de465b94c99c87aed8c6f12d6701](https://github.com/sagemath/sagetrac-mirror/commit/7cb6b3d19311de465b94c99c87aed8c6f12d6701)\n\n**Reviewer:** Fr\u00e9d\u00e9ric Chapoton\n\n**Author:** Erik Bray\n\nIssue created by migration from https://trac.sagemath.org/ticket/23785\n\n",
    "closed_at": "2017-09-11T13:47:05Z",
    "created_at": "2017-09-05T13:19:47Z",
    "labels": [
        "component: python3",
        "bug"
    ],
    "milestone": "https://github.com/sagemath/sage/milestones/sage-8.1",
    "title": "Fix sagelib setup.py install issues on Python 3",
    "type": "issue",
    "url": "https://github.com/sagemath/sage/issues/23785",
    "user": "https://github.com/embray"
}
```
This fixes a few issues, particularly with `clean_stale_files` on Python 3.  In particular, it prevents deletion of valid extension module and Python bytecode files, by better handling of `__pycache__` and the more complex filename extensions used in Python 3.

**CC:**  @fchapoton

**Branch/Commit:** [7cb6b3d19311de465b94c99c87aed8c6f12d6701](https://github.com/sagemath/sagetrac-mirror/commit/7cb6b3d19311de465b94c99c87aed8c6f12d6701)

**Reviewer:** Frédéric Chapoton

**Author:** Erik Bray

Issue created by migration from https://trac.sagemath.org/ticket/23785





---

archive/issue_comments_361337.json:
```json
{
    "body": "<a id='comment:1'></a>\n**Branch pushed to git repo; I updated commit sha1. This was a forced push.** **New commits:**\n<table><tr><td><a href=\"https://github.com/sagemath/sagetrac-mirror/commit/eb4c515a1e8d31855b70f905106bfea542739cf8\">eb4c515</a></td><td><code>Add a get_extensions() utility function for returning the system-defined Python module extensions--particularly important for getting the right bytecode and extension module filenames across platforms on Python 3</code></td></tr><tr><td><a href=\"https://github.com/sagemath/sagetrac-mirror/commit/8ef39d0ec8afcd6b329ee71ec073c1fbcb7c3f06\">8ef39d0</a></td><td><code>Use get_extensions() to get the correct file extensions here</code></td></tr><tr><td><a href=\"https://github.com/sagemath/sagetrac-mirror/commit/fa8f69ea466c5ed300756b7965d715aac5b495df\">fa8f69e</a></td><td><code>Fix installed_files_by_module to handle `__pycache__` on Python 3</code></td></tr><tr><td><a href=\"https://github.com/sagemath/sagetrac-mirror/commit/23f9fedc2bfe013a806d899d6ca8cda9c32bd2b8\">23f9fed</a></td><td><code>Fix _find_stale_files to stop removing .pyc files in the `__pycache__` directories.</code></td></tr></table>\n",
    "created_at": "2017-09-05T13:20:51Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361337",
    "user": "https://trac.sagemath.org/admin/accounts/users/git"
}
```

<a id='comment:1'></a>
**Branch pushed to git repo; I updated commit sha1. This was a forced push.** **New commits:**
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/eb4c515a1e8d31855b70f905106bfea542739cf8">eb4c515</a></td><td><code>Add a get_extensions() utility function for returning the system-defined Python module extensions--particularly important for getting the right bytecode and extension module filenames across platforms on Python 3</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/8ef39d0ec8afcd6b329ee71ec073c1fbcb7c3f06">8ef39d0</a></td><td><code>Use get_extensions() to get the correct file extensions here</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/fa8f69ea466c5ed300756b7965d715aac5b495df">fa8f69e</a></td><td><code>Fix installed_files_by_module to handle `__pycache__` on Python 3</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/23f9fedc2bfe013a806d899d6ca8cda9c32bd2b8">23f9fed</a></td><td><code>Fix _find_stale_files to stop removing .pyc files in the `__pycache__` directories.</code></td></tr></table>




---

archive/issue_comments_361338.json:
```json
{
    "body": "**Changing commit** from \"[12cb82f7cd2874de118f1b3c44ef2af8dcd8bac8](https://github.com/sagemath/sagetrac-mirror/commit/12cb82f7cd2874de118f1b3c44ef2af8dcd8bac8)\" to \"[23f9fedc2bfe013a806d899d6ca8cda9c32bd2b8](https://github.com/sagemath/sagetrac-mirror/commit/23f9fedc2bfe013a806d899d6ca8cda9c32bd2b8)\".",
    "created_at": "2017-09-05T13:20:51Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361338",
    "user": "https://trac.sagemath.org/admin/accounts/users/git"
}
```

**Changing commit** from "[12cb82f7cd2874de118f1b3c44ef2af8dcd8bac8](https://github.com/sagemath/sagetrac-mirror/commit/12cb82f7cd2874de118f1b3c44ef2af8dcd8bac8)" to "[23f9fedc2bfe013a806d899d6ca8cda9c32bd2b8](https://github.com/sagemath/sagetrac-mirror/commit/23f9fedc2bfe013a806d899d6ca8cda9c32bd2b8)".



---

archive/issue_events_213347.json:
```json
{
    "actor": "https://github.com/embray",
    "created_at": "2017-09-05T14:23:45Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "label": "needs review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/23785#event-213347"
}
```



---

archive/issue_comments_361339.json:
```json
{
    "body": "<a id='comment:3'></a>\nThis does not seem to work..",
    "created_at": "2017-09-06T06:12:14Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361339",
    "user": "https://github.com/fchapoton"
}
```

<a id='comment:3'></a>
This does not seem to work..



---

archive/issue_comments_361340.json:
```json
{
    "body": "<a id='comment:4'></a>\nReplying to [@fchapoton](#comment%3A3):\n> This does not seem to work..\n\nOkay--I appreciate you testing it but this is not the most helpful comment...  It works for me.  How does it not work for you?",
    "created_at": "2017-09-06T09:24:13Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361340",
    "user": "https://github.com/embray"
}
```

<a id='comment:4'></a>
Replying to [@fchapoton](#comment%3A3):
> This does not seem to work..

Okay--I appreciate you testing it but this is not the most helpful comment...  It works for me.  How does it not work for you?



---

archive/issue_comments_361341.json:
```json
{
    "body": "<a id='comment:5'></a>\nsee patchbot reports, sorry",
    "created_at": "2017-09-06T09:29:25Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361341",
    "user": "https://github.com/fchapoton"
}
```

<a id='comment:5'></a>
see patchbot reports, sorry



---

archive/issue_comments_361342.json:
```json
{
    "body": "<a id='comment:6'></a>\nAh, oops, thanks!  Sorry I misunderstood you.",
    "created_at": "2017-09-06T09:38:49Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361342",
    "user": "https://github.com/embray"
}
```

<a id='comment:6'></a>
Ah, oops, thanks!  Sorry I misunderstood you.



---

archive/issue_comments_361343.json:
```json
{
    "body": "<a id='comment:7'></a>\npatchbot says\n\n```\n[sagelib-8.1.beta4] Traceback (most recent call last):\n[sagelib-8.1.beta4]   File \"setup.py\", line 843, in <module>\n[sagelib-8.1.beta4]     SAGE_SRC, ['sage', 'sage_setup'])\n[sagelib-8.1.beta4]   File \"/home/kevin/sage-patchbot/src/sage_setup/find.py\", line 61, in find_python_sources\n[sagelib-8.1.beta4]     INIT_FILE = '__init__' + PYMOD_EXT\n[sagelib-8.1.beta4] TypeError: cannot concatenate 'str' and 'int' objects\n```",
    "created_at": "2017-09-06T19:24:09Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361343",
    "user": "https://github.com/fchapoton"
}
```

<a id='comment:7'></a>
patchbot says

```
[sagelib-8.1.beta4] Traceback (most recent call last):
[sagelib-8.1.beta4]   File "setup.py", line 843, in <module>
[sagelib-8.1.beta4]     SAGE_SRC, ['sage', 'sage_setup'])
[sagelib-8.1.beta4]   File "/home/kevin/sage-patchbot/src/sage_setup/find.py", line 61, in find_python_sources
[sagelib-8.1.beta4]     INIT_FILE = '__init__' + PYMOD_EXT
[sagelib-8.1.beta4] TypeError: cannot concatenate 'str' and 'int' objects
```



---

archive/issue_comments_361344.json:
```json
{
    "body": "**Changing commit** from \"[23f9fedc2bfe013a806d899d6ca8cda9c32bd2b8](https://github.com/sagemath/sagetrac-mirror/commit/23f9fedc2bfe013a806d899d6ca8cda9c32bd2b8)\" to \"[7cb6b3d19311de465b94c99c87aed8c6f12d6701](https://github.com/sagemath/sagetrac-mirror/commit/7cb6b3d19311de465b94c99c87aed8c6f12d6701)\".",
    "created_at": "2017-09-07T11:09:55Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361344",
    "user": "https://trac.sagemath.org/admin/accounts/users/git"
}
```

**Changing commit** from "[23f9fedc2bfe013a806d899d6ca8cda9c32bd2b8](https://github.com/sagemath/sagetrac-mirror/commit/23f9fedc2bfe013a806d899d6ca8cda9c32bd2b8)" to "[7cb6b3d19311de465b94c99c87aed8c6f12d6701](https://github.com/sagemath/sagetrac-mirror/commit/7cb6b3d19311de465b94c99c87aed8c6f12d6701)".



---

archive/issue_comments_361345.json:
```json
{
    "body": "<a id='comment:8'></a>\n**Branch pushed to git repo; I updated commit sha1.** **New commits:**\n<table><tr><td><a href=\"https://github.com/sagemath/sagetrac-mirror/commit/e86786d59d452b5aafff5af3b56ab9dd8adfaa4e\">e86786d</a></td><td><code>Fix typo in python2 implementation of get_extensions</code></td></tr><tr><td><a href=\"https://github.com/sagemath/sagetrac-mirror/commit/7cb6b3d19311de465b94c99c87aed8c6f12d6701\">7cb6b3d</a></td><td><code>Fix a cute bug with modules that are literally named 'module'</code></td></tr></table>\n",
    "created_at": "2017-09-07T11:09:55Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361345",
    "user": "https://trac.sagemath.org/admin/accounts/users/git"
}
```

<a id='comment:8'></a>
**Branch pushed to git repo; I updated commit sha1.** **New commits:**
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/e86786d59d452b5aafff5af3b56ab9dd8adfaa4e">e86786d</a></td><td><code>Fix typo in python2 implementation of get_extensions</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7cb6b3d19311de465b94c99c87aed8c6f12d6701">7cb6b3d</a></td><td><code>Fix a cute bug with modules that are literally named 'module'</code></td></tr></table>




---

archive/issue_events_213348.json:
```json
{
    "actor": "https://github.com/fchapoton",
    "created_at": "2017-09-07T17:36:21Z",
    "event": "unlabeled",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "label": "needs review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/23785#event-213348"
}
```



---

archive/issue_events_213349.json:
```json
{
    "actor": "https://github.com/fchapoton",
    "created_at": "2017-09-07T17:36:21Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "label": "positive review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/23785#event-213349"
}
```



---

archive/issue_comments_361346.json:
```json
{
    "body": "<a id='comment:9'></a>\nok, looks good enough. And bot is green.",
    "created_at": "2017-09-07T17:36:21Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361346",
    "user": "https://github.com/fchapoton"
}
```

<a id='comment:9'></a>
ok, looks good enough. And bot is green.



---

archive/issue_comments_361347.json:
```json
{
    "body": "**Reviewer:** Fr\u00e9d\u00e9ric Chapoton",
    "created_at": "2017-09-07T17:36:21Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361347",
    "user": "https://github.com/fchapoton"
}
```

**Reviewer:** Frédéric Chapoton



---

archive/issue_events_213350.json:
```json
{
    "actor": "https://github.com/vbraun",
    "created_at": "2017-09-11T13:47:05Z",
    "event": "unlabeled",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "label": "positive review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/23785#event-213350"
}
```



---

archive/issue_events_213351.json:
```json
{
    "actor": "https://github.com/vbraun",
    "commit_id": "9933f5e37cb88fedbb45d19df3e599369dca2183",
    "created_at": "2017-09-11T13:47:05Z",
    "event": "closed",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/23785#event-213351"
}
```



---

archive/issue_comments_361348.json:
```json
{
    "body": "**Changing branch** from \"[u/embray/python3/sagelib-build-fixes](https://github.com/sagemath/sagetrac-mirror/tree/u/embray/python3/sagelib-build-fixes)\" to \"[7cb6b3d19311de465b94c99c87aed8c6f12d6701](https://github.com/sagemath/sagetrac-mirror/commit/7cb6b3d19311de465b94c99c87aed8c6f12d6701)\".",
    "created_at": "2017-09-11T13:47:05Z",
    "issue": "https://github.com/sagemath/sage/issues/23785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/23785#issuecomment-361348",
    "user": "https://github.com/vbraun"
}
```

**Changing branch** from "[u/embray/python3/sagelib-build-fixes](https://github.com/sagemath/sagetrac-mirror/tree/u/embray/python3/sagelib-build-fixes)" to "[7cb6b3d19311de465b94c99c87aed8c6f12d6701](https://github.com/sagemath/sagetrac-mirror/commit/7cb6b3d19311de465b94c99c87aed8c6f12d6701)".
